packages/lib/hooks/useTypedQuery.ts

```diff
@@ -0,0 +1,59 @@
+import { useRouter } from "next/router";
+import { z } from "zod";
+
+export function useTypedQuery<T extends z.Schema>(schema: T) {
+  type InferedSchema = z.infer<typeof schema>;
+  type SchemaKeys = keyof InferedSchema;
+  const { query: unparsedQuery, ...router } = useRouter();
+  const parsedQuery = schema.parse(unparsedQuery) as InferedSchema;
+
+  // Set the query based on schema values
+  function setQuery<J extends SchemaKeys>(key: J, value: InferedSchema[J]) {
+    router.replace({ query: { ...parsedQuery, [key]: value } }, undefined, { shallow: true });
+  }
+
+  // Delete a key from the query
+  function removeByKey(key: SchemaKeys) {
+    const { [key]: _, ...newQuery } = parsedQuery;
+    router.replace({ query: newQuery }, undefined, { shallow: true });
+  }
+
+  // push item to existing key
+  function pushItemToKey<J extends SchemaKeys>(
+    key: J,
```

Comment:
You can filter this to only get the keys that rep arrays

Reply:
I honestly couldnt get this working - I was trying the ArrayKeys type above but was getting an error about not being able to set J to typeof<t> 

Maybe @zomars has more TS knowledge to debug this